Source/Core/Core/HW/DVDInterface.cpp

```diff
@@ -1169,138 +1167,297 @@ void FinishExecutingCommand(ReplyType reply_type, DIInterruptType interrupt_type
+
+  // Metroid Prime is a good example of a game that's sensitive to disc timing
+  // details; if there isn't enough latency in the right places, doors can open
+  // faster than on real hardware, and if there's too much latency in the wrong
+  // places, the video before the save-file select screen lags.
+
+  const u64 current_time = CoreTiming::GetTicks();
+
+  // Where the DVD read head is (usually parked at the end of the buffer,
+  // unless we've interrupted it mid-buffer-read).
+  u64 head_position;
+
+  // Compute the start (inclusive) and end (exclusive) of the buffer.
+  // If we fall within its bounds, we get DMA-speed reads.
+  u64 buffer_start, buffer_end;
+
+  if (SConfig::GetInstance().bFastDiscSpeed)
   {
-    // No buffer; just use the simple seek time + read time.
-    DEBUG_LOG(DVDINTERFACE, "Seeking %" PRId64 " bytes", s64(offset) - s64(s_last_read_offset));
-    ticks_until_completion = disk_read_duration;
-    s_last_read_time = current_time + ticks_until_completion;
+    // The SUDTR setting makes us act as if all reads are buffered
+    buffer_start = std::numeric_limits<u64>::min();
+    buffer_end = std::numeric_limits<u64>::max();
+    head_position = 0;
   }
   else
   {
-    // Possibly buffered; use the buffer if it saves time.
-    // It's not proven that the buffer actually behaves like this, but
-    // it appears to be a decent approximation.
-
-    // Time at which the buffer will contain the data we need.
-    u64 buffer_fill_time =
-        s_last_read_time +
-        CalculateRawDiscReadTime(s_last_read_offset, offset + length - s_last_read_offset);
-    // Number of ticks it takes to transfer the data from the buffer to memory.
-    u64 buffer_read_duration = length * (SystemTimers::GetTicksPerSecond() / BUFFER_TRANSFER_RATE);
-
-    if (current_time > buffer_fill_time)
+    if (s_read_buffer_start_time == s_read_buffer_end_time)
+    {
+      // No buffer
+      buffer_start = buffer_end = head_position = 0;
+    }
+    else
     {
-      DEBUG_LOG(DVDINTERFACE, "Fast buffer read at %" PRIx64, offset);
-      ticks_until_completion = buffer_read_duration;
-      s_last_read_time = buffer_fill_time;
+      buffer_start = s_read_buffer_end_offset > STREAMING_BUFFER_SIZE ?
+                         s_read_buffer_end_offset - STREAMING_BUFFER_SIZE :
+                         0;
+
+      DEBUG_LOG(DVDINTERFACE,
+                "Buffer: now=0x%" PRIx64 " start time=0x%" PRIx64 " end time=0x%" PRIx64,
+                current_time, s_read_buffer_start_time, s_read_buffer_end_time);
+
+      if (current_time >= s_read_buffer_end_time)
+      {
+        // Buffer is fully read
+        buffer_end = s_read_buffer_end_offset;
+      }
+      else
+      {
+        // The amount of data the buffer contains *right now*, rounded to a DVD ECC block.
+        buffer_end = s_read_buffer_start_offset +
+                     Common::AlignDown((current_time - s_read_buffer_start_time) *
```

Comment:
As far as I know, the only difference is that AlignDown works even when the second argument isn't a power of 2. Usage of ROUND_DOWN was replaced with AlignDown in PR #4482.